Make LANGUAGE section docs easier to navigate

mattias-p opened the pull request:
## Purpose

Today the `LANGUAGE section` documentation is quite long and if you're looking for something specific you may end up having to read all of it. This PR breaks it down into shorter sub-headings with clear titles.

## Context

I originally made these changes a good while ago, but I didn't get around to actually make a PR out of them until now.

## Changes

This PR moves around sections of text and inserts new headings, but it doesn't change any of the original words.

A list from the LANGUAGE.locale is integrated in to another one under `get_language_tags`.

Some text from LANGUAGE.locale and some from `get_language_tags` is integrated into the definition of the language tag data type definition.

## How to test this PR

This PR does not need testing as it only affects documentation.

matsduf commented on docs/API.md:
```diff
@@ -382,6 +382,17 @@ An array of *Profile names* in lower case. `"default"` is always included.
+
+If there are two `locale tags` that would give the same short
+`language tag` then that is excluded. E.g. "en_US en_UK" will
+only give "en_US" and "en_UK" as `language tags`.
```
Lines 385 to 394, in your proposal, should rather be integrated into the language tag section that this section already refers to. You have a reference to "locale tag" but no link.

mattias-p replied:
I fixed the link and tweaked the language a little bit.

I'm not sure I understand what your mean about integrating. But let me take this moment to explain why I moved these paragraphs here.

One way to to draw the line between the RPCAPI and the configuration is so that the RPCAPI is about "doing" and the configuration is about "being". From this perspective "generating" is a form of doing, and so it belongs in the RPCAPI. The configuration on the other hand just sits there quite inert concerning itself only with being valid and having a meaning. I believe this perspective is a good guiding principle that is fairly easy to understand and apply.

matsduf replied:
There are two sections in the API document on language tags. You have put the text under the `get_language_tags`method section. I think it is better to have all details under the "language tag" section and then referring to that.

mattias-p replied:
Oh, I see. Yeah. I'll make an update.

matsduf reviewed (changes requested):
It does improve the readability of the text.

pnax commented on docs/API.md:
```diff
@@ -382,6 +382,17 @@ An array of *Profile names* in lower case. `"default"` is always included.
+If "en_US" is the `locale tag` then "en" and "en_US" are the
+`language tags`.
+
+If there are two `locale tags` that would give the same short
+`language tag` then that is excluded. E.g. "en_US en_UK" will
+only give "en_US" and "en_UK" as `language tags`.
```
Would it be possible to rephrase this part a little bit ?
Maybe something like :
```
If there are two `locale tags` that would give the same short
`language tag`, this tag will then not be reported (or will be excluded)
```

matsduf replied:
```
If there are two `locale tags` that would give the same short
`language tag` then the short tag is excluded. E.g. "en_US en_UK" will
only give "en_US" and "en_UK" as `language tags` ("en" is excluded).
```

mattias-p replied:
How updating the paragraph above as well? E.g. like this:

```
Each `locale tag` in LANGUAGE.locale generates one or two `language tags`.
A long `language tag` identical to the `locale tag` is always generated.
A short `language tag` consisting of the first two letters of the `locale tag`
(usually the same as a language code) may also be generated.
The short tag is generated when there is only one `locale tag` that would
generate this particular short tag (i.e. so that it is unambiguous).

E.g. if LANGUAGE.locale is "en_US en_UK sv_SE", the generated `language tags`
are "en_US", "en_UK", "sv_SE" and "sv".
```

mattias-p replied:
This section is no more. I've integrated it into the language tag data type definition, and this resulted in a completely different wording.

pnax replied:
I think we lose information by removing it completely (see https://github.com/zonemaster/zonemaster-backend/pull/797#discussion_r655337591).

pnax reviewed (commented):
This is a good improvement, I like the new `Installation consideration` section.

mattias-p commented:
I added a couple of sub-headings to the *language tag* data type definition. Just like LANGUAGE.locale I thought it was a bit too long.

And I also moved around the *language tag* and *non-negative integer* data type definitions to make them comply with the lexicographic order.

pnax commented on docs/API.md:
```diff
@@ -156,6 +156,41 @@ This parameter is a string that are an IPv4 or IPv6. It's validated with the fol
+  first two characters of exactly one *locale tag* in LANGUAGE.locale.
+  (So that it is unambiguous which *locale tag* is matched.)
+
+E.g. if LANGUAGE.locale is "en_US en_UK sv_SE", all the valid *language tags*
```
Could you add a link on LANGUAGE.locale ?

mattias-p replied:
Not linking here follows Wikipedia's policy on repeated links (since LANGUAGE.locale is already linked a few lines above):
https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style/Linking#Duplicate_and_repeat_links

I like that guideline. It feels quite measured to me. Transferring their guideline to our context, I'd say "Language tag" passes for a glossary entry.

Thoughts on this?

pnax replied:
The idea is interesting. However I am wondering if this applies to documentation, as you would like to make it easier for readers to jump around.

pnax commented on docs/API.md:
````diff
@@ -391,16 +397,22 @@ Example request:
 ```
 
-Example response:
+Example response (as generated using the default configuration):
````
Have you considered the fact that this will imply that we need to update this part when we add a new language support ?

mattias-p replied:
@matsduf Would you object if I remove this?

pnax replied:
Not at all.

pnax commented on docs/Configuration.md:
```diff
@@ -195,17 +194,7 @@ English is the Zonemaster default language, but it can be blocked
-If "en_US" is the `locale tag` then "en" and "en_US" are the
-`language tags`.
-
-If there are two `locale tags` that would give the same short
-`language tag` then that is excluded. E.g. "en_US en_UK" will
-only give "en_US" and "en_UK" as `language tags`.
```
It looks like this paragraph has been removed with commit 0b0d5af2. I think we should keep somewhere the fact that we remove the short _language tag_ when two (or more) _locale tags_ result in the same short tag (ie: "en_US" and "en_GB" will not give "en").

mattias-p replied:
Isn't that clear enough from this?

> * If the *language tag* is a two-character string, it needs to match the
>   first two characters of exactly one *locale tag* in LANGUAGE.locale.
>   (So that it is unambiguous which *locale tag* is matched.)
>
>  E.g. if LANGUAGE.locale is "en_US en_UK sv_SE", all the valid *language tags*
>  are "en_US", "en_UK", "sv_SE" and "sv".

pnax replied:
From my point of view, both sentences describe the problem from a different perspective. One is starting from two _locale tags_ giving the same two-character _language tag_, the other one starts with one two-character _language tag_ which will match with the beginning of a _locale tag_. 
I think that to understand the output of `get_language_tags`, we should explain that we don't output the two-character _language tag_ when 2 (or more) _locale tags_ are resulting in the same short _language tag_.

mattias-p replied:
Sure, I'll make a stab at it. I'll adapt @matsduf's suggestion from above to make it work in its new context.

matsduf commented:
@pnax, can you re-review?

pnax commented on docs/API.md:
```diff
@@ -386,7 +386,11 @@ An array of *Profile names* in lower case. `"default"` is always included.
+
+> Note: If there are two [*locale tags*][LANGUAGE.locale] in [LANGUAGE.locale]
+> that would give the same [short language tag][Language tag] then the short tag
+> is excluded from the set of valid [*langauge tags*][Language tag].
```
small typo: langauge -> language

pnax reviewed (commented):
Thanks for the update. I spotted a typo.
